### PR TITLE
TokenBridgeRouter and CircleBridgeAdapter

### LINKED ISSUE #1202: LLI: Base Smart Contract
Smart contract changes for Liquidity Layer Interface

v0 stopgap solution is a Middleware Router that wraps Circle to start with and maybe Portal in time for ETHSF.

Some pseudo code is

```
    function dispatch(uint32 _destinationDomain, bytes32 _recipient, bytes message, address token, uint256 amount)
        external
    {
        token.transferFrom(msg.sender)

        address adapter = getAdapterForToken(_destinationDomain, token)
        adapter.transfer(_recipient, _destinationDomain, token, amount)

        dispatch()
    }

    function _handle(
        uint32 _origin,
        bytes32, // router sender
        bytes calldata _message
    ) internal override {
        (address recipient, bytes memory, address token, uint256 amount) = abi.decode(
            _message,
            (address, bytes, address, uint256)
        );

        address adapter = getAdapterForToken(_destinationDomain, token)
        adapter.checkTransfer()
        
        token.approve(recipient)
        recipient.handle(origin, sender, message, token, amount);
    }
  ```

### LINKED ISSUE #1203: LLI: Circle Smart Contract Adaptor
Example flow for Circle:

- `adapter.transfer`
- calls `depositForBurn` on the circle bridge with the remote router as the destination for the tokens
- returns the nonce and embeds it into the circle specific part of the hyperlane message
- `adapter.checkTransfer`
- parses out the nonce and checks on the circle bridge whether said nonce was processed
- `usedNonces` is the call to make, and the argument is the hash of the source domain and the nonce itself (see TG group)

### PR DESCRIPTION
**Pull request by nambrot**
### Description

This PR submits @tkporter POC of a adapter based `TokenBridgeRouter` with the first adapter being for the Circle token bridge. Deployment tooling and relayer POC come in subsequent PRs. There are probably plenty of nits to be had, but the goal is to get this ready for ETH SF, so review with that in mind.

### Related issues

- Fixes #1202  and fixes #1203 

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

Manual
Unit Tests

### CODE REVIEW

File: solidity/test/TokenBridgeRouter.t.sol [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+        originTokenBridgeRouter = new TokenBridgeRouter();
+        destinationTokenBridgeRouter = new TokenBridgeRouter();
+
+        hyperlaneTestHelperSetUp(originDomain, destinationDomain);
+
+        // TODO: set IGP?
+        originTokenBridgeRouter.initialize(
+            address(this),
+            address(originManager),
```
Comment (asaj) [inline solidity/test/TokenBridgeRouter.t.sol:44]: Favoring inheritance over composition makes this difficult to read. Consider instead composing HyperlaneTestEnvironment and having view calls on that (e.g. hyperlaneTestEnvironment.abacusConnectionManager(origin))

File: solidity/test/TokenBridgeRouter.t.sol [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            destinationDomain,
+            TypeCasts.addressToBytes32(address(recipient)),
+            messageBody,
+            address(token),
+            amount,
+            bridge
+        );
+    }
+
+    function testDispatchWithTokenTransfersMovesTokens() public {
```
Comment (asaj) [inline solidity/test/TokenBridgeRouter.t.sol:141]: Is this test missing st?

File: solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    ) internal override {
+        // Decode the message with metadata, "unwrapping" the user's message body
+        (
+            bytes32 _originalSender,
+            bytes32 _userRecipientAddress,
+            bytes memory _userMessageBody,
+            string memory _bridge,
+            uint256 _amount,
+            bytes memory _adapterData
+        ) = abi.decode(
```
[the remark this answers is not in the document]
Comment (nambrot) [reply solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol:96]: Remind me, what is the benefit of that again?

File: solidity/contracts/mock/MockTokenBridgeAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    function bridgeToken(
+        uint32 _destinationDomain,
+        bytes32 _recipientAddress,
+        address _token,
+        uint256 _amount
+    ) external override returns (bytes memory _adapterData) {
+        _destinationDomain;
+        _recipientAddress;
+        _token;
+        _amount;
```
[the remark this answers is not in the document]
Comment (nambrot) [reply solidity/contracts/mock/MockTokenBridgeAdapter.sol:26]: I think this was just a place holder to have the compiler not warn about unused args

File: solidity/test/TokenBridgeRouter.t.sol (hunk 2 above)
Comment (nambrot) [reply solidity/test/TokenBridgeRouter.t.sol:141]: I think originally it was just testing that the origin side does not fail

File: solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol:91]: It should be more gas efficient and contribute to a shallower stack. Plus it puts all the encoding/decoding logic in the same place so it's clear that it aligns

File: solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol [hunk 5]
```diff
@@ -89,33 +84,27 @@ contract TokenBridgeRouter is Router {
...
             );
 
         // Reverts if the adapter hasn't received the bridged tokens yet
-        (address _token, uint256 _sentAmount) = _adapter.sendBridgedTokens(
-            _origin,
-            address(_userRecipient),
-            _adapterData,
-            _amount
-        );
+        (address _token, uint256 _sentAmount) = getAdapter(_bridge)
```
Comment (asaj) [inline solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol:101]: consider "_receivedAmount"

File: solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol [hunk 6]
```diff
@@ -133,4 +122,14 @@ contract TokenBridgeRouter is Router {
         tokenBridgeAdapters[_bridge] = _adapter;
         emit TokenBridgeAdapterSet(_bridge, _adapter);
     }
+
+    function getAdapter(string memory _bridge)
```
Comment (asaj) [inline solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol:126]: nit: consider `_getAdapter`

File: solidity/test/HyperlaneTestEnvironment.sol [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import "../contracts/mock/MockOutbox.sol";
+import "../contracts/mock/MockInbox.sol";
+import "../contracts/AbacusConnectionManager.sol";
+
+contract HyperlaneTestEnvironment {
```
Comment (asaj) [inline solidity/test/HyperlaneTestEnvironment.sol:8]: Maybe it should be `MockHyperlaneEnvironment`? Wdyt?

File: solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (nambrot) [reply solidity/contracts/middleware/token-bridge/TokenBridgeRouter.sol:91]: I'll probably keep it as is, and let us move to it once we feel good for production

**Pull request merged**